### PR TITLE
Annotations documentation

### PR DESCRIPTION
**Pull request by meztez**
Brief description of the original problem and the approach behind your solution.

Document annotations, fix bugs and conflicting behavior along the way.

Remove openapi 2.0 annotations, update to annotations to openapi 3.0.3

PR task list:
- [x] `@assets` args order
- [x] Endpoint annotations
- [ ] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

### CODE REVIEW

File: R/plumb-globals.R [hunk 1]
```diff
@@ -32,37 +32,28 @@ plumbOneGlobal <- function(fields, argument){
...
          apiLicense={
+           if (grepl("^list\\(", def)) {
+             def <- eval(parse(text = def))
+           }
            fields$info$license <- def
          },
          apiVersion={
            fields$info$version <- def
          },
-         apiHost={
```
Comment (meztez) [inline R/plumb-globals.R:43]: Update to this function were made to match openapi 3.0.3 spec.

File: R/plumber.R [hunk 2]
```diff
@@ -118,7 +118,7 @@ Plumber <- R6Class(
           private$disable_run <- FALSE
         }, add = TRUE)
 
-        for (i in 1:length(private$parsed)){
+        for (i in seq_len(length(private$parsed))) {
```
Comment (meztez) [inline R/plumber.R:121]: While plumbing an empty file or a file without any block expression, this would create an error since the loop would go from 1:0 instead of not looping.

File: R/plumber.R [hunk 3]
```diff
@@ -817,7 +817,8 @@ Plumber <- R6Class(
       private$docs_info <- list(
         enabled = enabled,
         docs = docs,
-        args = list(...)
+        args = list(...),
+        has_not_been_set = FALSE
```
Comment (meztez) [inline R/plumber.R:821]: Saw that this was set during router initialization, I think it should be set to FALSE when `setDocs` is called. Otherwise it stays TRUE at all time.

File: R/pr.R [hunk 4]
```diff
@@ -499,21 +499,24 @@ pr_run <- function(pr,
...
 #' pr() %>%
-#'   pr_static("/path", "./my_folder/location") %>%
+#'   pr_static("./my_folder/location", "/path") %>%
 #'   pr_run()
 #' }
 #'
 #' @export
+# TODO: Validate if this makes sense?
+# Arguments path and direc are in reverse order compared to plumber annotation `@assets direc {path}`.
+# Default path is `/public` for annotation `@assets`
```
Comment (meztez) [inline R/pr.R:515]: What the comment says

File: R/plumb-globals.R [hunk 5]
```diff
@@ -32,37 +32,28 @@ plumbOneGlobal <- function(fields, argument){
...
-         },
-         apiProduces={
-           fields$produces <- strsplit(def, split="\\s+")[[1]]
-         },
          apiTag={
            tagMat <- stri_match(def, regex="^\\s*(\\w+)\\s+(\\S.+)\\s*$")
            name <- tagMat[1,2]
            description <- tagMat[1,3]
-           if(!is.null(fields$tags) && name %in% fields$tags$name) {
+           if(!is.null(fields$tags) && name %in% sapply(fields$tags, "[[", "name")) {
```
Comment (schloerke) [inline R/plumb-globals.R:53]: I try to avoid `sapply` at all costs due to the inconsistent simplifying behavior.  If we know it's a vector, we should use `vapply`. If it's a vector, but may not have a `name` in each item, then we can `unlist(lapply())`

```suggestion
           if(!is.null(fields$tags) && name %in% unlist(lapply(fields$tags, "[[", "name"))) {
```

File: R/plumber.R (hunk 3 above)
Comment (schloerke) [reply R/plumber.R:821]: I'm good with reversing the name / logic. Ex: ~ `vanilla = TRUE`

File: R/pr.R (hunk 4 above)
Comment (schloerke) [reply R/pr.R:515]: I think `@assets`'s order is wrong. I'd like to keep the original `pr_static()` setup.  

It might be nice to change the docs to not use `@assets` (due to the confusing order) and instead use....

```r
#' @plumber
function(pr) {
  pr %>%
    pr_static(path, direc)
}
```

File: R/plumber.R (hunk 3 above)
Comment (meztez) [reply R/plumber.R:821]: Let me rephrase that. `has_not_been_set` is not set anywhere else than initialization. So the value will always be what you set it to when you check it later for the purpose of lifecycle messaging. This boolean value should be reversed when you actually set the doc. This is what I was alluding to.

File: R/pr.R (hunk 4 above)
Comment (meztez) [reply R/pr.R:515]: My hypothesis is that the initial `@asset` args order was this way because `path` was optional. It default to `/public` when not provided. I could change the order and still making `path` optional by assuming:

One argument on line, set as directory
#* @asset directory
Two arguments on line, set as path and directory
#* @asset path directory

Is it ok?

Comment (meztez) [reply R/pr.R:515]: I will assume this is ok and go ahead with the changes.

Comment (meztez) [reply R/pr.R:515]: Also, what a sweeping job you did. Good job.

Comment (schloerke) [reply R/pr.R:515]: Unfortunately, we can not change the behavior of the tag `@assets` tag. This would definitely be breaking behavior. 

If anything we'd have to add a new tag (ex: `@static) and I'd like to pass on that for now.

When we add support for httpuv static paths (which don't support filters), we can make sure the new tag names are good.

File: vignettes/annotations.Rmd [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    pr_set_docs("swagger")
+}
+```
+
+##### Equivalent programmatic usage
+```{r, eval = FALSE, echo = TRUE}
+pr() %>%
+  pr_set_debug(TRUE) %>%
+  pr_set_docs("swagger")
+```
```
Comment (schloerke) [inline vignettes/annotations.Rmd:199]: This vignette is awesome. Thank you!

File: R/plumber.R (hunk 3 above)
Comment (schloerke) [reply R/plumber.R:821]: I'm up for changing the name / logic as long as we can tell the difference between the initialization and customized `docs_info`.  (I know my current implementation was a lil bit of a hack.)

**Review by schloerke — CHANGES_REQUESTED** [review]
If we can revert the `@assets` changes, I'm happy to merge! Thank you!

File: R/pr.R (hunk 4 above)
Comment (meztez) [reply R/pr.R:515]: Let's keep them in their original order then, even if mismatched.

**Review by schloerke — APPROVED** [review]

File: R/plumb-globals.R [hunk 7]
```diff
@@ -32,37 +32,28 @@ plumbOneGlobal <- function(fields, argument){
            fields$info$termsOfService <- def
          },
          apiContact={
+           if (grepl("^list\\(", def)) {
+             def <- eval(parse(text = def))
```
Comment (cpsievert) [inline R/plumb-globals.R:36]: This should probably have an `envir` coming in from the router

**Review by cpsievert — APPROVED** [review]
LGTM, pending suggested `envir` change

**Comment by schloerke** [comment]
@meztez Let me know when it's ready and I'll merge. (Please feel free to add a news entry for you adding the annotations vignette. Thank you!!)

**Comment by meztez** [comment]
> @meztez Let me know when it's ready and I'll merge. (Please feel free to add a news entry for you adding the annotations vignette. Thank you!!)

It is ready, let me finish investigate what github is complaining about.
